### PR TITLE
Eahw2281/save start date

### PR DESCRIPTION
**Pull request by IanFisherHO**
This update allows the user to set the start and end date associated with their shift start and end times. Notification functionality has also been added to inform them of successful updates.

### CODE REVIEW

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:184]: Is `timeEntry.startTime !== ''` just checking if a time entry already exists? 

In `EditShiftTimecard`, we use `const timeEntryExists = !!timeEntry?.startTime && timeEntry.startTime !== ''`. Wondering if it might be useful to create a similar constant to use in this file to make it easier to read?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:188]: This logic means that we only save the edited start date when the checkbox is checked, so if you close the component after editing, your new start date isn't saved. Is that the desired functionality?

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:536]: Could we put the checkbox tests into their own describe block?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:83]: This name suggests a generic handler but this is specifically for the checkbox.
We should update the name or place the arrow function directly in the handler param of the checkbox.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:183]: What is meant by a **local**Start/EndDate?

**Review by daleelR — APPROVED** [review]

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:146]: Agreed and removed

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:184]: Constant added for start and end date

File: src/utils/constants.js
Comment (IanFisherHO) [reply src/utils/constants.js:15]: Moved to it's own test utils class

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:83]: Updated to be relevant to the checkbox only functionality

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:183]: These are just local variables used so we can apply logic depending on whats populated between the inputted start date, the inputted start time and the timecard date

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx]: Moved to their own section

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx:573]: I think either this test should sit inside the "_given time entries are to be persisted_" describe block or it could sit along with the above test in a new describe block around checkbox related stuff.
There are a ton of tests in this suite and would be good to keep them organised.

File: src/utils/test-utils/testConstants.js
Comment (daleelR) [inline src/utils/test-utils/testConstants.js:1]: It seems like this object holds constants specifically for checkbox related input names, the variable name should reflect that. 
Or if it's for **any** input element names then i'd imagine we'd want the current properties within it to mention 'checkbox' otherwise it will be hard to distinguish when adding other unrelated ones.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:78]: Are tests to cover this functionality to come in the near future?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:184]: Variable no longer needed

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:148]: Moved to the calling function

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:169]: Wording has been updated

File: src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.spec.jsx]: Moved relevant tests into their own section

File: src/utils/test-utils/testConstants.js
Comment (IanFisherHO) [reply src/utils/test-utils/testConstants.js:1]: These are all test ID constants used for the date boxes at the moment, put here rather than mixing them up with the UI constants.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:76]: This is one of the focal points of next sprints spike

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:279]: The dates have been pulled out into their own component

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Dates are now in state and capture the changes regardless of the checkbox

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Variables have been renamed

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Conditions have been removed in some cases but are still required for end dates

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:88]: Just thinking that these get called when `summaryErrors` changes, which seems to be from line 240 when they are reset. It might be better to move this logic into another useEffect which gets called when `summaryMessages` changes

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:236]: Is it possible to submit without a start entry? Not sure if this if statement is needed

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:88]: Or a function that gets called after `setSummaryMessages` on line 176?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:236]: That is just there to stop notifications raising on new entries. It was easier to use that existing constant than creating a isNewRecord from it.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: yes makes sense for start date ones only. Thanks

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Ah I meant on the function parameters themselves, so that it's clear what kind of information the function expects.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:83]: cool, should we consider if we need to track this untested functionality somehow, e.g. a TODO comment or it written in a ticket somewhere before the PR gets merged...
Please feel free to resolve this.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx:91]: Created second useEffect and split the code out

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: I tried this but caused quite a lot of failures. I'm reluctant to change this as part of this PR given the level of changes implemented.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Sorry, updated as suggeted

Comment (daleelR) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: can't see the change, but no worries, wont hold up the PR. Resolving..

**Comment by daleelR** [comment]
When I save a time entry, then change the startdate and save, the notification banner show 'Invalid Date' for the end date.
image

- [ ] Address Issue

**Comment by daleelR** [comment]
When I save a time entry and then edit the start date only, it frequently gives me a server error:
image

- [ ] Address Issue

File: src/components/common/form/message-item-types/DoubleDateMoved.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DoubleDateMoved.jsx:16]: Could these be in an object instead of array?

File: src/components/common/form/message-item-types/DoubleDateMoved.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DoubleDateMoved.jsx:48]: Could we use PropTypes.shape with the start and end date instead of PropTypes.array?

File: src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx:26]: Do we need to click the link after the expect statement?

File: src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx:8]: This implies it's rendering a SingleDateMoved component for the DoubleDateMoved test - looks like the import is wrong

File: src/components/common/form/message-item-types/SingleDateMoved.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/SingleDateMoved.jsx:1]: Do we need to have a separate file for this? Could we have some logic around whether the end date is passed in to decide which message to display?

File: src/components/common/form/message-item/MessageItem.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-item/MessageItem.spec.jsx:8]: Should there also be a test for doubleDateMoved?

File: src/components/common/form/message-summary/MessageSummary.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-summary/MessageSummary.spec.jsx:9]: I know we don't have multiple messages right now, but should there be a test for the scenario for multiple summary messages being displayed at once to test the map function?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:211]: Are these 00:00 and 23:59? Could this be simplified?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:221]: I'm not 100% sure on the second part of the logic for these variables, might be one to have a quick call on?

File: src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx
Comment (joannafawl) [reply src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx:26]: I realised after going through all the files that this is probably to clear the summary messages - could we reset them before each test or add a comment to this line instead?

File: src/components/common/form/message-item/MessageItem.jsx
Comment (daleelR) [inline src/components/common/form/message-item/MessageItem.jsx:26]: This doesn't feel very React like, and also it results in us having duplication between `SingleDateMoved` and `DoubleDateMoved`.
What we're trying to achieve looks like it could be solved through React composition:
https://reactjs.org/docs/composition-vs-inheritance.html
Based on React's suggestion for a typical inheritance situation it looks like we could have a base like `DateMoved` component which contains whats currently in `SingleDateMoved` and then any additional ui e.g. what's in `DoubleDateMoved` could be  passed in as `children`. 
That way you could remove the need to rely on passing the string name of a component to render it and the duplication across both components.
Happy to discuss, maybe it's not an issue but probs worth while considering.

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [inline src/components/timecard/edit-shift-hours/EditShiftHours.jsx:236]: If we go with the composition related comment, instead of passing in a template to be associated with a component name, we could just pass a boolean for `singleDateMoved`, then in the `MessageItem` if `singleDateMoved` is false we can additionally render the `DoubleDateMoved` stuff

File: src/components/common/form/message-item-types/DoubleDateMoved.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DoubleDateMoved.jsx]: Updated to be an object

File: src/components/common/form/message-item-types/DoubleDateMoved.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DoubleDateMoved.jsx]: I've change the prop type to object based on the previous comment

File: src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx]: Updated typos

File: src/components/common/form/message-item/MessageItem.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item/MessageItem.spec.jsx]: Test added for double date scenario

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (daleelR) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Resolving after discussion based on other related comment.

File: src/components/common/form/message-item-types/DatesMoved.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DatesMoved.spec.jsx:3]: Minor but could we change this so the import is DatesMoved?

Comment (joannafawl) [reply src/components/common/form/message-item-types/DatesMoved.spec.jsx:3]: And similar with the describe block below

File: src/components/common/form/message-item/MessageItem.spec.jsx
Comment (joannafawl) [inline src/components/common/form/message-item/MessageItem.spec.jsx:17]: What is this keys prop? It isn't a prop in the component

Comment (IanFisherHO) [reply src/components/common/form/message-item/MessageItem.spec.jsx]: It's used by the message summary to determine what order the messages should be displayed in if there's more than one

File: src/components/common/form/message-item-types/DatesMoved.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DatesMoved.spec.jsx]: Updated typo to be inline with component name

File: src/components/common/form/message-item/MessageItem.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item/MessageItem.jsx]: Code updated to move the keys to the constants file

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Simplified dates to use less variables and easier to read comparisons

File: src/components/common/form/message-summary/MessageSummary.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-summary/MessageSummary.spec.jsx:9]: I've added a test to show we can handle multiple messages being input, but since we can only have one of each key and only handle one key, it won't show the second message

File: src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DoubleDateMoved.spec.jsx]: Updated the tests to move the link click to it's own test and make sure it's calling the expected function

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: I've tried various scenarios but can't make this fail.

File: src/components/common/form/message-item-types/DatesMoved.jsx
Comment (joannafawl) [inline src/components/common/form/message-item-types/DatesMoved.jsx:6]: Now that we only have one prop passed in, does it make sense to pass in startDate and endDate directly?

File: src/components/common/form/message-item-types/DatesMoved.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DatesMoved.spec.jsx:14]: Agreed, I've removed them

File: src/components/common/form/message-item-types/DatesMoved.spec.jsx
Comment (joannafawl) [reply src/components/common/form/message-item-types/DatesMoved.spec.jsx]: Could you also change the describe block?

File: src/pages/timecard/Timecard.spec.jsx
Comment (joannafawl) [inline src/pages/timecard/Timecard.spec.jsx:321]: Is it possible to test that the message summary is cleared when clicking previous/next day?

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (IanFisherHO) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: I set up a shift with the start and end on 17/03/23. I then moved the start date to be 16/03/23 and no notification was displayed, which is correct.

File: src/pages/timecard/Timecard.spec.jsx
Comment (IanFisherHO) [reply src/pages/timecard/Timecard.spec.jsx:321]: Extra test added

File: src/components/common/form/message-item-types/DatesMoved.spec.jsx
Comment (IanFisherHO) [reply src/components/common/form/message-item-types/DatesMoved.spec.jsx]: Updated

File: src/components/timecard/edit-shift-hours/EditShiftHours.jsx
Comment (joannafawl) [reply src/components/timecard/edit-shift-hours/EditShiftHours.jsx]: Resolving as we think this is covered by EAHW-2279 - will double check functionality with BAs then

**Review by joannafawl — APPROVED** [review]

**Review by daleelR — APPROVED** [review]

**Pull request merged**
